rstudio/rstudio
# Add i18n for menu headings and commands

### Intent

See #9083 for more context on i18n efforts.

This PR enables i18n for the menu headings and command headings/descriptions that are defined in the `Commands.cmd.xml` file.  Shortcuts in that file have not been addressed here

### Approach

#### Implementation

* 	Code for the menu (eg: `File->New File->...`) and cmd (`Undo`, `Redo`, ...) items in `Commands.cmd.xml` is created using GWT generators at compile time (`CommandsBundleGenerator.java` and `MenuEmitter.java`).  
* 	To enable i18n, the generators have been modified to create `.java` files that invoke GWT i18n constants interfaces instead of hard-coded strings for all user-facing text in menus/cmds
*	A helper script `src/gwt/tools/i18n-helpers/commands.cmd/xml/commands_xml_to_i18n.py` is added automate generating the i18n interfaces and an English properties file from the `Commands.cmd.xml` file.  Similar to the generators, it creates a string entry for each user-facing text.  The products of running this script (interfaces `CmdConstants.java`/`MenuConstants.java` and corresponding English properties files `CmdConstants_en.properties`/`MenuConstants_en.properties`) are committed.  More instructions for use are included in `src/gwt/tools/README.md`

#### Developer Workflow for i18n with `Commands.cmd.xml`

For any changes to `Commands.cmd.xml` that affect user-facing text (adding/removing menus and commands, changing labels/descriptions of commands, etc.), developers will:
*	Modify `Commands.cmd.xml`
*	Rerun `commands_xml_to_i18n.py` to regenerate java interfaces and English properties files
*	(if supporting additional languages) make comparable changes to other language properties files

This workflow is similar to that for syncing the RStudio preferences and state accessors for the R session (C++) and the front end (GWT Java) based on JSON schema files using `/src/gwt/tools/generate-prefs.R`

### Automated Tests

No automated tests were added.  I am not sure if this content is covered in any automated tests, but if it is they will need to be internationalized before adding any new translations

### QA Notes

Manual QA can be done by:
*   loading in `devmode`
*   browsing to rstudio at `localhost:8787/` or `localhost:8787/?locale=en` to see standard English text (these should look identical, and be the same as the current `main` branch)
*   creating a "dev" locale and viewing it by:
  *  following instructions in `/src/gwt/tools/i18n-helpers/README.md` and using `/src/gwt/tools/i18n-helpers/create_dev_local.sh` to create a "dev" locale that has English text with an `@` symbol prepended to all i18n texts
  * browsing to `localhost:8787/?locale=dev` to see the "dev" locale with `@` sumbol like shown below

![image](https://user-images.githubusercontent.com/48125859/122588969-b264fa80-d02d-11eb-8ebc-1549b0db84ed.png)

### Checklist


### Other

* there are two duplicate entries for a `importDatasetFromSAV` cmd in `Commands.cmd.xml`.  They differ slightly, but it appears this duplication may be a bug?  I have commented this out as duplicates make i18n ambiguous and am hoping this was a mistake, but appreciate feedback if this is wrong
* Sorry, the `commands_xml_to_i18n.py` is in python.  I'm sure it could have been in R, but my R skills are poor and it was python or nothing :)
